### ISSUE TITLE
threadpool hang due to unlimited hill climbing

### ISSUE THREAD

**Issue by StanZanarotti** [mono/mono#17833]
The Mono threadpool can deadlock under certain workload patterns, due to a hill climbing algorithm that increases the max_working counter when too many workers get created per second. Even though the workers exit due to inactivity, the counter does not decrease. Once the max_working counter reaches limit_worker_max (set to 100 * number of CPUS * threads_per_cpu), the threadpool is easily vulnerable to deadlock, even with a small number of actual threads in use (< 30). This can cause random hangs in programs that make heavy use of the thread pool in a periodic fashion.

Background: We have a production (batch) processing system written in C#. For the past two years, it's been randomly hanging on Mono after many days of uptime, without any hangs happening in .NET.

I've traced it down to the threadpool deadlock -- creating a periodic load of creating many workers at once followed by a minute or more of quiescence increases the variable worker.counters._.max_working in mono/metadata/threadpool-worker-default.c. Once that counter matches worker.limit_worker_max, monitor_thread() keeps looping without unparking or creating a thread, even though the actual number of threads/workers is small. If the existing threads are all waiting on work that needs a new thread, then a deadlock occurs.

I've created a test program that illustrates this on all versions of Mono from 5.10.1.20 to 6.4.0.198. It queues 31 work items, waits for all of them to be active, tells them to go away, and then waits 60 seconds. Within 28 minutes on a 2 CPU system, this program will hang the threadpool (~14 minutes per cpu). This test program does not hang under .NET.

I've also enclosed an excerpt of threadpool debug output of a run of our production service (only the hill climbing lines), showing how max number of threads reaches 200 on a 2 CPU system. Not shown is the fact that the program only uses 15-30 threads during this entire time.

Also attached is gdb output of monitor_thread::worker during a hang.

## Steps to Reproduce

1.  Compile and run threadpool-deadlock.cs.
2. Wait until it hangs (time dependent on # of CPUs)

## On which platforms did you notice this

[ ] macOS
[X] Linux
[ ] Windows

**Version Used**:

5.10.1.20
5.18.0.240
6.4.0.198

Attachments:
[hill.txt](https://github.com/mono/mono/files/3861730/hill.txt)
[gdb.txt](https://github.com/mono/mono/files/3861734/gdb.txt)
[threadpool-deadlock.txt](https://github.com/mono/mono/files/3861738/threadpool-deadlock.txt)

## Stacktrace

```
Please paste the stack trace here if available.
```

**Comment by jaykrell** [comment]
Does the code confuse `working` and `max_working`?
I mean, is `max_working` even so important, vs. current `working`?

I guess, I mean, what if you remove these lines:

```
		COUNTER_ATOMIC (counter, {
			if (counter._.max_working >= worker.limit_worker_max) {
				limit_worker_max_reached = TRUE;
				break;
			}
			counter._.max_working ++;
		});
```

**Comment by jaykrell** [comment]
I need to read this more..

I guess, the 200 threads to stay created and will respond to work, at least.
They don't exit.

**Comment by jaykrell** [comment]
But ultimately if you have 200 work items all depending on "each other", you'll get the same deadlock, right? Should there really be a maximum number of threads?
We should probably test that too.

**Comment by StanZanarotti** [comment]
Yes, that would produce the same deadlock, although at that point you would have created 200 threads. In this case, the limit is 200, but Mono hangs with only 20 threads in use.

**Comment by jaykrell** [comment]
I understand it hangs with fewer -- your description was pretty clear. :)

**Comment by jaykrell** [comment]
Hm. Is `hill_climbing_update` supposed to sometimes compute `max_workers`? 
Via:
 `heuristic_adjust`
 `heuristic_notify_work_completed` 
 `mono_threadpool_worker_notify_completed`
 `heuristic_should_adjust`
?


Still, it feels like `worker_thread` exits without enough accounting.
and worse `monitor_thread` adds to `max_workers` without trying unpark.

**Comment by marek-safar** [comment]
@lambdageek please have a look

**Comment by lambdageek** [comment]
Brief update: I'm currently focused on trying to understand our TP code and [the corresponding code in CoreCLR](https://github.com/dotnet/coreclr/blob/5f5f9606a6eef78256536f667c2df7b2495a8c3d/src/vm/win32threadpool.cpp#L4329-L4342) to understand why they don't have the same problem.

**Comment by StanZanarotti** [comment]
I can give you a couple insights from looking back at code history:

Testing against Mono 4.2.1 (the first version copied from Microsoft as threadpool-ms.c), the threads stuck around parked in the pool. Draining the thread pool after a minute of idleness was introduced later in #4113 (threadpool-worker-default.c:500).

Also in #4113, there was a limit of 10 threads created per second, which caused the max_workers counter to get incremented every time that happened. I tried to remove the increment logic for this situation, but the max_workers counter still got incremented in other circumstances, so just removing this limit is not sufficient to avoid the problem.

**Comment by lambdageek** [comment]
What I find interesting is that in Mono we don't apparently have any way to lower max_working - it only ever increases, except if the heuristic decides to adjust it down.  On the other hand in CoreCLR, when a worker times out, [they adjust max_working down to the number of active threads](https://github.com/dotnet/runtime/blob/bf88f146412d1e2d41779422337184a802f186c9/src/coreclr/src/vm/win32threadpool.cpp#L2156-L2164).

I'm going to investigate adding something similar at the end of `worker_thread` in case `worker_park` returned that it timed out.  (I'm not sure why we don't have a `num_active` counter anymore, but I think `num_working + num_parked` should work, too)


I don't think 4117 added the timeout.  It was there in threadpool-ms.c previously.  I don't understand the downstream effects of the 10 threads / sec limit - I guess without it we hit the "starvation" case in the monitor thread less frequently? 

In any case, I'm going to see what happens if I allow timed out threads to lower max_working.

**Comment by lambdageek** [comment]
Fixed on master and 2019-12 (#17927) and 2019-10 (#17944) now.

**Comment by HydraOrc** [comment]
Similar issue happened to me in my project.
Was running websocket-sharp library with mono, and it wasn't handling too many connections. Sometimes 12, sometimes 28, etc.
Had to switch this library to another one, specifically vtorola.websockets, and now it can handle much more clients with mono.
